include/Contact.php

```diff
@@ -508,72 +507,100 @@ function contacts_not_grouped($uid,$start = 0,$count = 0) {
+ * directly return the found contact id.
+ *
+ * Second, we probe the provided $url wether it's http://server.tld/profile or
+ * nick@server.tld. We quit if we can't get any info back.
+ *
+ * Third, we create the contact record if it doesn't exist
+ *
+ * Fourth, we update the existing record with the new data (avatar, alias, nick)
+ * if there's any updates
+ *
  * @param string $url Contact URL
- * @param integer $uid The user id for the contact
+ * @param integer $uid The user id for the contact (0 = public contact)
  * @param boolean $no_update Don't update the contact
  *
  * @return integer Contact ID
  */
 function get_contact($url, $uid = 0, $no_update = false) {
-	require_once("include/Scrape.php");
-
 	logger("Get contact data for url ".$url." and user ".$uid." - ".App::callstack(), LOGGER_DEBUG);;
 
 	$data = array();
-	$contactid = 0;
-
-	// is it an address in the format user@server.tld?
-	/// @todo use gcontact and/or the addr field for a lookup
-	if (!strstr($url, "http") OR strstr($url, "@")) {
-		$data = probe_url($url);
-		$url = $data["url"];
-		if ($url == "")
-			return 0;
-	}
+	$contact_id = 0;
 
-	$contact = q("SELECT `id`, `avatar-date` FROM `contact` WHERE `nurl` = '%s' AND `uid` = %d ORDER BY `id` LIMIT 2",
+	// We first try the nurl (http://server.tld/nick), most common case
+	$contacts = q("SELECT `id`, `avatar-date` FROM `contact`
+					WHERE `nurl` = '%s'
+					AND `uid` = %d",
 			dbesc(normalise_link($url)),
 			intval($uid));
 
-	if (!$contact)
-		$contact = q("SELECT `id`, `avatar-date` FROM `contact` WHERE `alias` IN ('%s', '%s') AND `uid` = %d ORDER BY `id` LIMIT 1",
-				dbesc($url),
-				dbesc(normalise_link($url)),
-				intval($uid));
 
-	if ($contact) {
-		$contactid = $contact[0]["id"];
+	// Then the addr (nick@server.tld)
+	if (! dbm::is_result($contacts)) {
+		$contacts = q("SELECT `id`, `avatar-date` FROM `contact`
+					WHERE `addr` = '%s'
+					AND `uid` = %d",
+			dbesc($url),
+			intval($uid));
+	}
+
+	// Then the alias (which could be anything)
+	if (! dbm::is_result($contacts)) {
+		$contacts = q("SELECT `id`, `avatar-date` FROM `contact`
+					WHERE `alias` IN ('%s', '%s')
+					AND `uid` = %d",
+			dbesc($url),
+			dbesc(normalise_link($url)),
+			intval($uid));
+	}
+
+	if (dbm::is_result($contacts)) {
+		$contact_id = $contacts[0]["id"];
 
 		// Update the contact every 7 days
-		$update_photo = ($contact[0]['avatar-date'] < datetime_convert('','','now -7 days'));
-		//$update_photo = ($contact[0]['avatar-date'] < datetime_convert('','','now -12 hours'));
+		$update_photo = ($contacts[0]['avatar-date'] < datetime_convert('','','now -7 days'));
 
 		if (!$update_photo OR $no_update) {
-			return($contactid);
+			return $contact_id;
 		}
-	} elseif ($uid != 0)
+	} elseif ($uid != 0) {
+		// Non-existing user-specific contact, exiting
+		return 0;
+	}
+
+	require_once('include/Probe.php');
+	$data = Probe::uri($url);
+	if (!$data['url']) {
```

Comment:
I just saw that you changed the behavior here. Please have a look at the current code. When the given url doesn't belong to Friendica, Diaspora or OStatus it tries to fetch some data from the gcontact table. Only if this fails as well, we are quitting.